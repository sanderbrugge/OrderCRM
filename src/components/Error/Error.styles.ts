import { StyleSheet } from "react-native";
import { colors } from "../../styles/base";

const ErrorStyles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "row",
    alignItems: "center",
    height: 40,
    borderColor: colors.yellow,
    borderWidth: 2,
    margin: 10,
    borderRadius: 40,
    backgroundColor: colors.yellow
  },
  errorIcon: {
    color: colors.white,
    fontSize: 20,
    marginLeft: 10,
    marginRight: 10
  },
  errorMessage: {
    color: colors.white,
    fontSize: 18
  }
});

export default ErrorStyles;
