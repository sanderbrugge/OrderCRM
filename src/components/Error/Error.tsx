import * as React from "react";
import { View, Text } from "react-native";
import FontAwesome, { Icons } from "react-native-fontawesome";
import ErrorStyles from "./Error.styles";

interface IProps {
  message: string;
}

const Error: React.FC<IProps> = ({ message }) => {
  return (
    <View style={ErrorStyles.container}>
      <FontAwesome style={ErrorStyles.errorIcon}>{Icons.bug}</FontAwesome>
      <Text style={ErrorStyles.errorMessage}>{message}</Text>
    </View>
  );
};

export default Error;
