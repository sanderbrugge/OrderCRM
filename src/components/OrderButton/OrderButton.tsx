import * as React from "react";
import { TouchableOpacity, Text } from "react-native";
import { colors, baseColorWithOpacity } from "../../styles/base";
import FontAwesome from "react-native-fontawesome";
import OrderButtonStyles from "./OrderButton.styles";

interface IProps {
  title?: string;
  color?: string;
  icon?: string;
  disabled?: boolean;
  onClick: () => void;
}

const OrderButton: React.FC<IProps> = ({
  title,
  color,
  disabled,
  icon,
  onClick
}) => {
  const backgroundColor = disabled
    ? baseColorWithOpacity(colors.blue, "80")
    : color
    ? color
    : colors.blue;

  return (
    <TouchableOpacity
      disabled={disabled}
      onPress={onClick}
      style={[
        OrderButtonStyles.container,
        {
          backgroundColor
        }
      ]}
    >
      <Text style={OrderButtonStyles.buttonText}>
        {icon ? (
          <FontAwesome>
            {icon}
            {title ? ` ${title}` : null}
          </FontAwesome>
        ) : (
          <Text>{title}</Text>
        )}
      </Text>
    </TouchableOpacity>
  );
};

export default OrderButton;
