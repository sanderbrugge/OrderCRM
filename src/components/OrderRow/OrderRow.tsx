import * as React from "react";
import { Order } from "../../api/Order/order";
import { TouchableOpacity, View, Text } from "react-native";
import { OrderStyles } from "./Order.styles";
import FontAwesome, { Icons } from "react-native-fontawesome";

interface IProps {
  order: Order;
  customerName: string;
  onSelect: () => void;
}

/**
 * Dumb-component that is only used to render out the info
 *
 * @param param0 The order to display, the customer's name and the on row select handler
 */
const OrderRow: React.FC<IProps> = ({ order, customerName, onSelect }) => (
  <TouchableOpacity onPress={onSelect}>
    <View style={OrderStyles.container}>
      <View style={OrderStyles.thumbnail}>
        <FontAwesome style={OrderStyles.icon}>{Icons.shoppingCart}</FontAwesome>
        <View style={OrderStyles.itemCount}>
          <Text style={OrderStyles.itemCountText}>{order.items.length}</Text>
        </View>
      </View>
      <View style={OrderStyles.detailsContainer}>
        <Text style={OrderStyles.customerDetailText}>{customerName}</Text>
        <Text style={OrderStyles.orderDetailText}>#{order.id}</Text>
        <Text style={OrderStyles.detailTextPrice}>${order.total}</Text>
      </View>
    </View>
  </TouchableOpacity>
);

export default OrderRow;
