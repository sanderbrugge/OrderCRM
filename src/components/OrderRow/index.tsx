import OrderRowContainer from "./OrderRowContainer";

export default OrderRowContainer;
