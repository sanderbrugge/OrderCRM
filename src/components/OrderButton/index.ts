import OrderButton from "./OrderButton";

export default OrderButton;
