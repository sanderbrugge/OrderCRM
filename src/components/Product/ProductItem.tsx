import * as React from "react";
import { Product } from "../../api/Product/product";
import { View, Text } from "react-native";
import OrderButton from "../OrderButton";
import ProductStyles from "./Product.styles";
import { Item } from "../../api/Order/order";

interface IProps {
  product: Product;
  orderId: string;
  orderProduct: (orderId: string, item: Item) => void;
}

function mapProductToOrderItem(product: Product): Item {
  return {
    "product-id": product.id,
    quantity: "1",
    "unit-price": product.price,
    total: product.price
  };
}

const ProductItem: React.FC<IProps> = ({ product, orderId, orderProduct }) => {
  const item = React.useMemo(() => mapProductToOrderItem(product), [product]);
  const onOrderProductClick = React.useCallback(() => {
    orderProduct(orderId, item);
  }, [product]);

  return (
    <View style={ProductStyles.container}>
      <View style={ProductStyles.productInfoContainer}>
        <Text style={ProductStyles.productDescription}>
          {product.description}
        </Text>
        <Text style={ProductStyles.prodcutId}>{product.id}</Text>
      </View>
      <View style={ProductStyles.buttonContainer}>
        <OrderButton
          title={`$${product.price}`}
          onClick={onOrderProductClick}
        />
      </View>
    </View>
  );
};

export default ProductItem;
