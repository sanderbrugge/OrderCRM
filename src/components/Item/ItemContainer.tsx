import * as React from "react";
import { Item } from "../../api/Order/order";
import { View, Text } from "react-native";
import ItemStyles from "./Item.styles";
import ItemInformation from "./ItemInformation";
import { getProductById } from "../../api/Product/product.services";
import { Product } from "../../api/Product/product";

interface IProps {
  item: Item;
}

/**
 * Overview of an item
 * @param param0 the item to display
 */
const ItemContainer: React.FC<IProps> = ({ item }) => {
  const [product, setProduct] = React.useState<Product | undefined>(undefined);
  const productName = product ? product.description : item["product-id"];
  React.useEffect(() => {
    const fetchProductData = async () => {
      try {
        const data = await getProductById(item["product-id"]);
        setProduct(data);
      } catch (error) {
        console.log(error);
        // can be used for additional error handling
      }
    };
    fetchProductData();
  }, []);
  return (
    <View style={ItemStyles.container}>
      <Text style={ItemStyles.productName}>{productName}</Text>
      <Text style={ItemStyles.productId}>{item["product-id"]}</Text>
      <View style={ItemStyles.overviewContainer}>
        <ItemInformation title="QUANTITY" info={item.quantity} />
        <ItemInformation title="UNIT PRICE" info={item["unit-price"]} />
        <ItemInformation title="TOTAL" info={item.total} />
      </View>
    </View>
  );
};

export default ItemContainer;
