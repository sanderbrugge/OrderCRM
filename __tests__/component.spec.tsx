import "react-native";
import * as React from "react";
import ProductItem from "../src/components/Product";
import Error from "../src/components/Error";
import ItemContainer from "../src/components/Item";
import OrderButton from "../src/components/OrderButton";
import OrderRow from "../src/components/OrderRow";
import renderer from "react-test-renderer";

test("Error Component renders correctly", () => {
  const tree = renderer.create(<Error message="Test error" />).toJSON();

  expect(tree).toMatchSnapshot();
});

test("ItemContainer renders correctly", () => {
  const item = {
    "product-id": "B102",
    quantity: "10",
    "unit-price": "4.99",
    total: "49.90"
  };
  const tree = renderer.create(<ItemContainer item={item} />).toJSON();

  expect(tree).toMatchSnapshot();
});

test("OrderButton Component renders correctly", () => {
  const tree = renderer
    .create(
      <OrderButton
        title="testtitle"
        color={"#FFF"}
        icon={"testicon"}
        onClick={jest.fn()}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("OrderRow Component renders correctly", async () => {
  jest.mock("react-navigation", () => ({
    withNavigation: (component: React.FC) => component
  }));

  const order = {
    id: "2",
    "customer-id": "2",
    items: [
      {
        "product-id": "B102",
        quantity: "5",
        "unit-price": "4.99",
        total: "24.95"
      }
    ],
    total: "24.95"
  };

  const mockProps = {
    navigation: { navigate: jest.fn() },
    order
  };

  const tree = await renderer.create(<OrderRow {...mockProps} />).toJSON();

  expect(tree).toMatchSnapshot();
});

test("ProductItem renders correctly", () => {
  const product = {
    id: "A102",
    description: "Electric screwdriver",
    category: "1",
    price: "49.50"
  };

  const orderId = "1";
  const orderProduct = jest.fn();

  const tree = renderer
    .create(
      <ProductItem
        product={product}
        orderId={orderId}
        orderProduct={orderProduct}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
