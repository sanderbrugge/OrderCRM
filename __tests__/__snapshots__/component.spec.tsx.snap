// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f5c300",
      "borderColor": "#f5c300",
      "borderRadius": 40,
      "borderWidth": 2,
      "flex": 1,
      "flexDirection": "row",
      "height": 40,
      "margin": 10,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Object {
          "color": undefined,
        },
        Object {
          "color": "#ffffff",
          "fontSize": 20,
          "marginLeft": 10,
          "marginRight": 10,
        },
        Object {
          "fontFamily": "FontAwesome5FreeSolid",
        },
      ]
    }
  >
    
  </Text>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "fontSize": 18,
      }
    }
  >
    Test error
  </Text>
</View>
`;

exports[`ItemContainer renders correctly 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "#a5a5a530",
      "borderBottomWidth": 1,
      "flex": 1,
      "marginLeft": 20,
      "marginTop": 20,
      "paddingBottom": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 4,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 18,
            "fontWeight": "600",
            "marginBottom": 2,
          }
        }
      >
        B102
      </Text>
      <Text
        style={
          Object {
            "color": "#a5a5a5",
            "fontSize": 14,
            "fontStyle": "italic",
          }
        }
      >
        B102
      </Text>
    </View>
    <View
      style={
        Object {
          "marginRight": 20,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#DC143C",
            "borderRadius": 200,
            "justifyContent": "center",
            "margin": 5,
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "alignItems": "center",
              "color": "#ffffff",
              "fontSize": 16,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "color": undefined,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome5FreeSolid",
                },
              ]
            }
          >
            
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": 2,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#999999",
            "fontSize": 14,
          }
        }
      >
        QUANTITY
      </Text>
      <Text
        style={
          Object {
            "fontWeight": "500",
            "marginTop": 10,
          }
        }
      >
        10
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#999999",
            "fontSize": 14,
          }
        }
      >
        UNIT PRICE
      </Text>
      <Text
        style={
          Object {
            "fontWeight": "500",
            "marginTop": 10,
          }
        }
      >
        $ 4.99
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#999999",
            "fontSize": 14,
          }
        }
      >
        TOTAL
      </Text>
      <Text
        style={
          Object {
            "fontWeight": "500",
            "marginTop": 10,
          }
        }
      >
        $ 49.90
      </Text>
    </View>
  </View>
</View>
`;

exports[`OrderButton Component renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFF",
      "borderRadius": 200,
      "justifyContent": "center",
      "margin": 5,
      "opacity": 1,
      "padding": 10,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "alignItems": "center",
        "color": "#ffffff",
        "fontSize": 16,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
          },
          Object {
            "color": undefined,
          },
          undefined,
          Object {
            "fontFamily": "FontAwesome5FreeSolid",
          },
        ]
      }
    >
      testicon
       testtitle
    </Text>
  </Text>
</View>
`;

exports[`OrderRow Component renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#a5a5a530",
        "borderBottomWidth": 1,
        "flex": 1,
        "flexDirection": "row",
        "marginBottom": 10,
        "paddingBottom": 5,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#a5a5a5",
          "borderColor": "#a5a5a5",
          "borderWidth": 1,
          "height": 70,
          "justifyContent": "center",
          "margin": 5,
          "width": 70,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "color": undefined,
            },
            Object {
              "color": "#ffffff",
              "fontSize": 30,
            },
            Object {
              "fontFamily": "FontAwesome5FreeSolid",
            },
          ]
        }
      >
        
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#008000",
            "borderRadius": 20,
            "flex": 1,
            "height": 20,
            "justifyContent": "center",
            "position": "absolute",
            "right": 8,
            "top": 12,
            "width": 20,
            "zIndex": 999,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
            }
          }
        >
          1
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 8,
        }
      }
    >
      <Text
        style={
          Object {
            "flex": 1,
            "fontSize": 18,
            "fontWeight": "500",
            "marginBottom": 5,
          }
        }
      >
        Customer id: 2
      </Text>
      <Text
        style={
          Object {
            "color": "#a5a5a5",
            "flex": 1,
            "fontSize": 14,
            "fontWeight": "300",
          }
        }
      >
        # 2
      </Text>
      <Text
        style={
          Object {
            "alignSelf": "flex-end",
            "color": "#008000",
            "flex": 1,
            "fontSize": 15,
            "fontWeight": "600",
          }
        }
      >
        $ 24.95
      </Text>
    </View>
  </View>
</View>
`;

exports[`ProductItem renders correctly 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "#a5a5a530",
      "borderBottomWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "height": 70,
    }
  }
>
  <View
    style={
      Object {
        "flex": 3,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "500",
          "marginBottom": 5,
          "marginTop": 10,
        }
      }
    >
      Electric screwdriver
    </Text>
    <Text
      style={
        Object {
          "color": "#a5a5a5",
          "fontSize": 16,
          "fontStyle": "italic",
          "fontWeight": "300",
        }
      }
    >
      A102
    </Text>
  </View>
  <View
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#4169E1",
          "borderRadius": 200,
          "justifyContent": "center",
          "margin": 5,
          "opacity": 1,
          "padding": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "alignItems": "center",
            "color": "#ffffff",
            "fontSize": 16,
            "justifyContent": "center",
          }
        }
      >
        <Text>
          $ 49.50
        </Text>
      </Text>
    </View>
  </View>
</View>
`;
